HMRC-1685: Handle and Propagate State Parameter

amberstarlight opened the pull request:
### Jira link

[HMRC-1685](https://transformuk.atlassian.net/browse/HMRC-1685)

### What?

I have added/removed/altered:

- Updated Session and Passwordless controllers to propagate state parameter if supplied by calling service

### Why?

I am doing this because:

- In the dev hub we will generate a state parameter that gets included with redirects to the Identity service (https://github.com/trade-tariff/trade-tariff-dev-hub/pull/65)
- We need to send this back when we redirect back to the calling service so that it can ensure the request was made in the service.

amberstarlight committed:
HMRC-1685: Propagate state parameter

amberstarlight committed:
HMRC-1685: Improve state handling

amberstarlight committed:
Fix inverted if

amberstarlight committed:
HMRC-1685: Add test cases

willfish reviewed (dismissed):
Hey Amber, a few things I spotted going through this:

 Redis shouldn't be needed here. The `state` parameter originates from the calling service (dev-hub) and is already stored in *its* session. Identity just needs to pass it through as a query param on the redirect back. It doesn't need to persist it at all. Adding Redis as a new infrastructure dependency for what's essentially a passthrough value is overkill and adds operational complexity we don't need.

Using email as the cache key is a problem. In the passwordless controller you're doing `Rails.cache.write(email, session[:state])` - if the same user opens two tabs or two users somehow share an email, they'll clobber each other's state value. It's not session-scoped.

`Rails.cache.delete` after `redirect_to` won't run. In the callback action, the `Rails.cache.delete(email)` call sits after `redirect_to`, so it'll never execute - the response is already sent. The cache entry will just sit there until the 1-day TTL expires.

Dead branching in `check_session`. Both sides of the `if session[:state].nil?` check redirect to `uri` - the conditional doesn't do anything different:
```ruby
if session[:state].nil?
  redirect_to uri, allow_other_host: true and return
end

redirect_to uri, allow_other_host: true and return
```
This is duplicated for both the `:valid` and `:expired` paths.

The simpler approach for identity would be: receive `state` as a param, stash it in the session (which you're already doing on the sessions controller `index` action), and append it to the redirect URL when sending the user back. No Redis, no cache, no email keys.

willfish reviewed (commented):
Weird github comment error

amberstarlight committed:
HMRC-1685: Revise implementation

We do not need to cache the session state value in the Identity service;
the service only needs to capture the value if it recieves one and append
it to any redirects back to the calling services.

- Removed Redis
- Removed dead branching in `check_session`; URI is already set so we
  don't need to take conditional action a second time

amberstarlight committed:
Merge branch 'main' into hmrc-1685-state

willfish reviewed (approved):
Approving to enable stale PR cleanup workflows.
